File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> subSet = new ArrayList<>();
        helper(candidates, target, subSet, new ArrayList<>(), 0);
        return subSet;
    }

    public void helper(int[] candidates, int target, List<List<Integer>> subSet, List<Integer> temp, int start) {
        Arrays.sort(candidates);
        if (target == 0) {
            subSet.add(new ArrayList<>(temp));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (candidates[i] > target) {
                continue;
            }
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;
            }
            temp.add(candidates[i]);
            helper(candidates, target - candidates[i], subSet, temp, i + 1);
            temp.remove(temp.size() - 1);
        }
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        candidates.sort()
        def helper(i,arr,target):
            if target == 0:
                ans.append(arr[:])
                return
            for indx in range(i,len(candidates)):
                if indx > i  and candidates[indx] == candidates[indx-1]:
                    continue
                if candidates[i] > target:
                    break
                arr.append(candidates[indx])
                helper(indx+1,arr,target-candidates[indx])  
                arr.pop()
        helper(0,[],target)
        return ans
        